### PR TITLE
UK Base World: population, regions, and interactions.

### PR DESCRIPTION
**Pull request by mdhillman**
This PR brings in changes to the UK Base World visualisation as well as corresponding updates to the TWA-VF to support the new functionality.

To test this PR, the steps for running the visualisation should be followed and the visualisation tested. In addition, using the example Mapbox visualisation to test the new IRI placeholders and the "interactions" parameter would be worthwhile.

Note that this PR is related to #889; if it's possible to get that PR done in a short time scale (a week), I would recommend handling that first then this one.

Overview of changes:

* In UK Base World visualisation:
  * Added country and county borders for UK
  * Added UK population count
  * Added geospatial SQL to calculate population within 1KM of each power plant
  * Can toggle visual 1KM radius for selected power plant
  * Updated legend and attributions
  
* In TWA-VF:
  * Added an "interactions" layer parameter.
  * Added the ability to use "[SELECTED-IRI]" & "[HOVERED-IRI]" placeholders in Mapbox expressions.
  * Added system to set callbacks for feature selection, unselection, and layer tree change events. 
  * Fixed minor issue with loading metadata from selected raster layers in Cesium.

> [!WARNING]
> Note that just before this PR is merged, the version of the TWA-VF will need the "-SNAPSHOT" token removed. In fact, the automated checks *should* fail until this is done.

### CODE REVIEW

**Review by cmclscourtney — CHANGES_REQUESTED** [review]
- [ ] Make clearer which "counties" dataset is required.
- [ ] Clarify what is meant by high and low population in the legend.

File: web/uk-base-world/inputs/config/uploader/sql/dukes_2023_pop.sql
Comment (Ushcode) [inline web/uk-base-world/inputs/config/uploader/sql/dukes_2023_pop.sql:1]: committed a suggestion to use CTE for circles instead of creating and dropping

**Review by Ushcode — CHANGES_REQUESTED** [review]

**Comment by Ushcode** [comment]
### Inside viz

- Show / Hide 1KM radius button doesn't seem to work
- FIA not working for power lines or generation sites

```2023-11-16 11:55:06.340 (FeatureInfoAgent)[http-nio-8080-exec-10] INFO - Request received at: 2023-11-16 11:55:06.0340
2023-11-16 11:55:06.340 (AgentCaller)[http-nio-8080-exec-10] INFO - accept = application/json, text/javascript, */*; q=0.01
2023-11-16 11:55:06.340 (FeatureInfoAgent)[http-nio-8080-exec-10] INFO - Detected request to get meta and times series data.
2023-11-16 11:55:06.340 (QueryManager)[http-nio-8080-exec-10] INFO - Incoming IRI is: http://theworldavatar.com/ontology/ontopower/OntoPower.owl#ng_overhead_lines/62111
2023-11-16 11:55:06.340 (QueryManager)[http-nio-8080-exec-10] INFO - No incoming enforced endpoint, will attempt federation.
2023-11-16 11:55:06.341 (ClassHandler)[http-nio-8080-exec-10] DEBUG - Running non-federated class determination query.
2023-11-16 11:55:06.392 (ClassHandler)[http-nio-8080-exec-10] ERROR - Running class determination query has thrown an exception!
java.lang.IllegalStateException: Class determination query has failed to return any results, cannot continue!
        at com.cmclinnovations.featureinfo.core.ClassHandler.determineClassMatches(ClassHandler.java:86) [classes/:?]
        at com.cmclinnovations.featureinfo.QueryManager.determineClasses(QueryManager.java:170) [classes/:?]
        at com.cmclinnovations.featureinfo.QueryManager.processRequest(QueryManager.java:132) [classes/:?]
        at com.cmclinnovations.featureinfo.FeatureInfoAgent.getRoute(FeatureInfoAgent.java:207) [classes/:?]
        at com.cmclinnovations.featureinfo.FeatureInfoAgent.doGet(FeatureInfoAgent.java:131) [classes/:?]
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:670) [servlet-api.jar:4.0.FR]
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:779) [servlet-api.jar:4.0.FR]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:227) [catalina.jar:9.0.69]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) [catalina.jar:9.0.69]
        at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53) [tomcat-websocket.jar:9.0.69]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189) [catalina.jar:9.0.69]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) [catalina.jar:9.0.69]
        at org.apache.logging.log4j.web.Log4jServletFilter.doFilter(Log4jServletFilter.java:71) [log4j-web-2.17.1.jar:2.17.1]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189) [catalina.jar:9.0.69]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) [catalina.jar:9.0.69]
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:177) [catalina.jar:9.0.69]
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:97) [catalina.jar:9.0.69]
        at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:541) [catalina.jar:9.0.69]
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:135) [catalina.jar:9.0.69]
        at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92) [catalina.jar:9.0.69]
        at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:687) [catalina.jar:9.0.69]
        at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:78) [catalina.jar:9.0.69]
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:360) [catalina.jar:9.0.69]
        at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:399) [tomcat-coyote.jar:9.0.69]
        at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65) [tomcat-coyote.jar:9.0.69]
        at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:891) [tomcat-coyote.jar:9.0.69]
        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1784) [tomcat-coyote.jar:9.0.69]
        at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49) [tomcat-coyote.jar:9.0.69]
        at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191) [tomcat-util.jar:9.0.69]
        at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659) [tomcat-util.jar:9.0.69]
        at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) [tomcat-util.jar:9.0.69]
        at java.lang.Thread.run(Thread.java:833) [?:?]
2023-11-16 11:55:06.393 (FeatureInfoAgent)[http-nio-8080-exec-10] ERROR - Could not run /get route.
java.lang.NullPointerException: Cannot invoke "java.util.List.forEach(java.util.function.Consumer)" because "classMatches" is null
        at com.cmclinnovations.featureinfo.core.meta.MetaHandler.getData(MetaHandler.java:85) ~[classes/:?]
        at com.cmclinnovations.featureinfo.QueryManager.getMeta(QueryManager.java:196) ~[classes/:?]
        at com.cmclinnovations.featureinfo.QueryManager.processRequest(QueryManager.java:138) ~[classes/:?]
        at com.cmclinnovations.featureinfo.FeatureInfoAgent.getRoute(FeatureInfoAgent.java:207) ~[classes/:?]
        at com.cmclinnovations.featureinfo.FeatureInfoAgent.doGet(FeatureInfoAgent.java:131) [classes/:?]
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:670) [servlet-api.jar:4.0.FR]
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:779) [servlet-api.jar:4.0.FR]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:227) [catalina.jar:9.0.69]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) [catalina.jar:9.0.69]
        at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53) [tomcat-websocket.jar:9.0.69]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189) [catalina.jar:9.0.69]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) [catalina.jar:9.0.69]
        at org.apache.logging.log4j.web.Log4jServletFilter.doFilter(Log4jServletFilter.java:71) [log4j-web-2.17.1.jar:2.17.1]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189) [catalina.jar:9.0.69]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) [catalina.jar:9.0.69]
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:177) [catalina.jar:9.0.69]
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:97) [catalina.jar:9.0.69]
        at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:541) [catalina.jar:9.0.69]
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:135) [catalina.jar:9.0.69]
        at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92) [catalina.jar:9.0.69]
        at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:687) [catalina.jar:9.0.69]
        at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:78) [catalina.jar:9.0.69]
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:360) [catalina.jar:9.0.69]
        at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:399) [tomcat-coyote.jar:9.0.69]
        at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65) [tomcat-coyote.jar:9.0.69]
        at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:891) [tomcat-coyote.jar:9.0.69]
        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1784) [tomcat-coyote.jar:9.0.69]
        at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49) [tomcat-coyote.jar:9.0.69]
        at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191) [tomcat-util.jar:9.0.69]
        at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659) [tomcat-util.jar:9.0.69]
        at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) [tomcat-util.jar:9.0.69]
        at java.lang.Thread.run(Thread.java:833) [?:?]
2023-11-16 11:55:06.395 (FeatureInfoAgent)[http-nio-8080-exec-10] INFO - Call finished, response object's writer has been flushed.
```

File: web/uk-base-world/inputs/config/uploader/sql/dukes_2023_pop.sql
Comment (mdhillman) [reply web/uk-base-world/inputs/config/uploader/sql/dukes_2023_pop.sql:1]: This change caused SQL errors during upload, have reverted for now. We can look into making the  query more efficient in future updates (focussed on functionality for this PR).

**Comment by mdhillman** [comment]
- Issue with toggling 1Km radius has been resolved.
- I cannot reproduce any issues with the FIA queries. However, calls to it for the overhead lines do take a long time (up to 2 minutes).

**Review by Ushcode — APPROVED** [review]

**Review by Ushcode — APPROVED** [review]

**Review by Ushcode — APPROVED** [review]

**Review by Ushcode — APPROVED** [review]

**Review by Ushcode — APPROVED** [review]

**Comment by mdhillman** [comment]
For the record, @cmclscourtney's requested changes have been make, but for whatever reason GitHub will not let me mark them as such.

**Pull request merged**